rfcs/0038-unified-device-target-and-memory-scope-planning.md

```diff
@@ -0,0 +1,244 @@
+2. Though TVM's `Device` abstraction (an alias for `dlpack`'s `DLDevice`) is a pair of a
+   `DLDeviceType` and an arbitrary 'device id', TVM does not consistently plumb the device id
+   through annotations, passes and operators.  Thus currently we cannot use 'device id' to
+   distinguish, eg, two CPUs in the same system.
+3. Upcoming work requires us to distinguish and propagate memory scopes for data at the Relay
+   level. (See also [RFC #9](https://github.com/apache/tvm-rfcs/blob/main/rfcs/0009_Unified_Static_Memory_Planning.md)
+   which has a similar need for memory scope propagation at the TIR level). This is an identical
+   problem to propagating devices, and it seems most natural to simply combine targets, devices and
+   memory scopes into a single 'target of device planing' rather than implementing a whole new pass.
+4. Device planning currently has no machinery to hoist adjacent expressions which share the same device
+   into their own Relay `Function`. For all our executors except VM that's unnecessary anyway since
+   all Relay expressions left over after lowering are interpreted by the runtime. However for AOT we
+   have to compile *all* Relay code for a particular target. Note the BOYC machinery does support this,
+   but for the purposes of redirecting the compilation flow entirely. We need a middle ground.
+5. The BYOC flow is not connected to the `Target` machinery in any way.
+6. The BYOC annotate/partition flow is very similar to the device annotate/rewrite flow. For comparison:
+
+   | Feature               | Device Planning            | BYOC                                            |
+   | --------------------- | -------------------------- | ----------------------------------------------- |
+   | Source of annotations | `on_device`, `device_copy` | `FTVMAnnotateTarget`, `MergeComposite`+patterns |
+   | Target of planning    | DLDeviceType               | Toolchain name                                  |
+   | Propagation           | Unification based          | Ad-hoc                                          |
+   | Relay support         | Full                       | First-order, no ADTs                            |
+   | Delimiting            | insert `device_copy`       | insert `compiler_begin`, `compiler_end`         |
+   | Multiple per expr     | No                         | Yes (though always picks first)                 |
+   | Hoists into functions | No                         | Yes                                             |
+   | Customized heuristics | No                         | No                                              |
+
+   Taking the 'upper bound' of the two implementations seems ideal, especially to address issues 4 (limitation
+   of device planning) and 5 (limitation of BYOC) above.
+
+Our proposal is:
+1. We introduce a new FFI-friendly class to represent a *S*torage or *E*xecution *Scope*:
+
+   ```
+   class SEScope {
+     DLDeviceType device_type;
+     int virtual_device_id;
```

Comment:
i think this should be a String name which makes sense to the user. Doing this is helpful for a couple other reasons besides the compilation UI:
- In generated source code, it's possible to refer to the device by name. In particular, the embedded C API would like to have this for the conglomerate tvm_device_t struct.
- In systems with multiple e.g. CPUs, using an index here then implies some ordering (e.g. littlest CPU to biggest). It's better to make the assignment of ID to CPU capability more explicit

Finally, using a name would simplify the heterogeneous Target.

However, this is a bit of a lift. I do feel strongly we should get to this world. If it's not something that makes sense to do now, we could also revisit after or concurrent with USMP.